node: Use new asynchronous IPCache API for Manager (v2)

christarazi opened the pull request:
On top of https://github.com/cilium/cilium/pull/22086 with additional fixes.

Related: https://github.com/cilium/cilium/issues/21142

gandro committed:
ipcache: Add TunnelPeer/EncryptKey metadata types

This commit adds new types for specifying encrypt key and tunnel peer as
IPCache metadata in the new asynchronous API. In contrast to other
metadata such as labels, the tunnel IP or encryption key cannot be
merged from other sources. It is assumed that only one source provides
this information. If multiple sources provide a tunnel peer or
encryption key, a log warning is emitted and only one of the provided
peers/keys is picked.

This will allow us to move the node/manager package to use this new API
in a subsequent commit, as previously the only way to associate this
data with a node IP was using the old synchronous API.

Co-authored-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

gandro committed:
nodemanager: Move to async IPCache API

This commit moves the node/manager package to use the new asynchronous
IPCache API. Instead of directly performing Upserts and Delete on the
various node IPs (InternalIP, ExternalIP, HealthIPs etc), we now
associate each node IP with the corresponding labels. The CEW identity
is now also determined by the node's labels, rather than its numeric
identity.

This also fixes an issue where concurrent use of the synchronous and
asynchronous API would lead to the encryption key for the kube-apiserver
node being lost (c.f. cilium/cilium#19318).

While we are at it, change the test to use netip types instead of
net.IP.

Co-authored-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

thorn3r commented on pkg/ipcache/types.go:
```diff
@@ -176,5 +229,33 @@ func (s prefixInfo) logConflicts(scopedLog *logrus.Entry) {
 		}
+
+		if info.tunnelPeer.IsValid() {
+			if tunnelPeer.IsValid() {
```
won't this always evaluate true after the 1st iteration with a valid `info.tunnelPeer`? or is the expectation here that only a single `info.tunnelPeer.IsValid()` in the given range?

christarazi replied:
cc @gandro

gandro replied:
Yes, there is the expectation that there is only a single info.tunnelPeer.IsValid() in the given range. The code you highlighted is the one reporting conflicts (via logging), the actual extraction of the (hopefully only) tunnelPeer happens here:

https://github.com/cilium/cilium/blob/70087b48029c397f412c22b7007299a67a1d90f0/pkg/ipcache/types.go#L155

Due to the asynchronous nature of the API, we cannot report this to the caller when it provides a conflicting value, so we check for conflicts and emit warnings instead.

gandro replied:
To give a bit more context: For metadata like labels, we can merge them if multiple source provide different labels. But for tunnelIPs, there can only be one IP (since the datapath doesn't support multiple tunnel endpoints), and we therefore assume that this one IP is only coming from a single source.

I we could in theory introduce an equality check, and allow multiple source to specify the tunnel IP, as long as all source agree, but that will due to the eventually consistent nature of our data sources, that will likely lead to temporary states where different sources report a different tunnel IP - thus we currently require that only one source provides a tunnel IP.

thorn3r replied:
that makes sense, thanks for the explanation! @gandro

ldelossa reviewed (approved):
I don't have the full context for the async change, but from a pure code logic standpoint this looks good to me.

jrajahalme reviewed (approved):
Not related to this PR, but would it be possible to always have a nodeIP in the IP cache? Now that is only done when tunneling, encryption, or host firewall is enabled, of when executing on an external workload. In other words, would there be any harm always populating the "tunnel IP" in IP cache entries, and maybe renaming it as "node IP"?

joestringer commented on pkg/ipcache/ipcache.go:
```diff
@@ -497,6 +497,15 @@ func (ipc *IPCache) RemoveMetadata(prefix netip.Prefix, resource ipcacheTypes.Re
 }
 
+// RemoveAllMetadata removes all metadata associated with this resource
+func (ipc *IPCache) RemoveAllMetadata(prefix netip.Prefix, resource ipcacheTypes.ResourceID) {
```
Could you explain a bit more deeply the rationale behind removing all? Does this mean that if two users of the new API insert metadata associations for different resources, then some other user comes along and uses this API, it will delete all of the metadata associations from the other user? That sounds counter-intuitive to the goals of having the ipcache itself receive desired state coming in and then handling the merging / resolution internally.

joestringer replied:
I'm thinking about a use case like this:

- Policy engine creates an association that IP 192.0.2.3 needs a rich identity corresponding to a policy statement
- Endpoint watcher observes that 192.0.2.3 corresponds to kube-apiserver and adds an association corresponding to that resource
- Node manager observes that 192.0.2.3 is a remote node in the cluster.
- At some point 192.0.2.3 gets removed from the cluster, now `RemoveAllMetadata()` deletes the policy association so policy to that remote peer no longer works until... maybe the user restarts the agent?

gandro replied:
That's a fair question. I think my original intent was to keep the behavior as close to the previous code as possible (and the previous code issues a delete-all too). I think you make an excellent argument though, we should not delete everything here. I'll attempt to replace it with a regular remove labels.

gandro replied:
Replaced `RemoveAllMetadata` with a function that only removes what was upserted during NodeUpdate.

gandro committed:
node: Simplify IPToNetPrefix conversion

Since cilium/cilium#21183 it is no longer necessary to call `To4` before
invoking `IPToNetPrefix`.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
node: Extract tunnel IP and encrypt key conditions

This commit does not contain any functional changes. It will be used in
a subsequent commit to invoke the same logic when removing the tunnel IP
and encrypt key.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

joestringer commented on pkg/ipcache/metadata_test.go:
```diff
@@ -234,6 +235,35 @@ func TestOverrideIdentity(t *testing.T) {
+	ip, key = IPIdentityCache.getHostIPCache(worldPrefix.String())
+	assert.True(t, net.ParseIP("192.168.1.200").Equal(ip))
+	assert.Equal(t, uint8(0), key)
+}
```
I presume we should also check that `upsertLocked()` on the same IP with differing info should return an error?

joestringer replied:
(Basically just a request for additional test coverage based on the assumption that we can't get the same info from multiple sources --- though some of the discussion below seems to suggest that we may actually receive node updates for example from both k8s and etcd in some scenarios). Minor nit, not blocking the PR. We could always expand the unit test coverage after this PR.

christarazi replied:
Do you mean to assert that an upsert with a weaker source (that contains a different tunnel IP) does not change the original ipcache entry?

christarazi replied:
I tried messing around with this and I'm not getting the behavior I'm expecting. I applied this diff here (ignore the lines that switch from `worldPrefix` to `inClusterPrefix`):

<details><summary>Diff</summary>
<p>

```diff
diff --git a/pkg/ipcache/metadata_test.go b/pkg/ipcache/metadata_test.go
index b93cec72eb..e9db8b1d07 100644
--- a/pkg/ipcache/metadata_test.go
+++ b/pkg/ipcache/metadata_test.go
@@ -258,25 +258,37 @@ func TestUpsertMetadataTunnelPeerAndEncryptKey(t *testing.T) {
 
 	ctx := context.Background()
 
-	IPIdentityCache.metadata.upsertLocked(worldPrefix, source.CustomResource, "node-uid",
+	IPIdentityCache.metadata.upsertLocked(inClusterPrefix, source.CustomResource, "node-uid",
 		types.TunnelPeer(netip.MustParseAddr("192.168.1.100")),
 		types.EncryptKey(7))
-	remaining, err := IPIdentityCache.InjectLabels(ctx, []netip.Prefix{worldPrefix})
+	remaining, err := IPIdentityCache.InjectLabels(ctx, []netip.Prefix{inClusterPrefix})
 	assert.NoError(t, err)
 	assert.Len(t, remaining, 0)
 
-	ip, key := IPIdentityCache.getHostIPCache(worldPrefix.String())
+	ip, key := IPIdentityCache.getHostIPCache(inClusterPrefix.String())
 	assert.True(t, net.ParseIP("192.168.1.100").Equal(ip))
 	assert.Equal(t, uint8(7), key)
 
-	IPIdentityCache.metadata.upsertLocked(worldPrefix, source.CustomResource, "node-uid",
+	// Assert that an entry with a weaker source (and from a different
+	// resource) should fail or at least does not overwrite the existing
+	// (stronger) ipcache entry.
+	IPIdentityCache.metadata.upsertLocked(inClusterPrefix, source.Generated, "generated-uid",
+		types.TunnelPeer(netip.MustParseAddr("192.168.1.101")),
+		types.EncryptKey(7))
+	_, err = IPIdentityCache.InjectLabels(ctx, []netip.Prefix{inClusterPrefix})
+	assert.Error(t, err)
+	ip, key = IPIdentityCache.getHostIPCache(inClusterPrefix.String())
+	assert.True(t, net.ParseIP("192.168.1.100").Equal(ip))
+	assert.Equal(t, uint8(7), key)
+
+	IPIdentityCache.metadata.upsertLocked(inClusterPrefix, source.CustomResource, "node-uid",
 		types.TunnelPeer(netip.MustParseAddr("192.168.1.200")))
-	IPIdentityCache.metadata.remove(worldPrefix, "node-uid", types.EncryptKey(7))
-	remaining, err = IPIdentityCache.InjectLabels(ctx, []netip.Prefix{worldPrefix})
+	IPIdentityCache.metadata.remove(inClusterPrefix, "node-uid", types.EncryptKey(7))
+	remaining, err = IPIdentityCache.InjectLabels(ctx, []netip.Prefix{inClusterPrefix})
 	assert.NoError(t, err)
 	assert.Len(t, remaining, 0)
 
-	ip, key = IPIdentityCache.getHostIPCache(worldPrefix.String())
+	ip, key = IPIdentityCache.getHostIPCache(inClusterPrefix.String())
 	assert.True(t, net.ParseIP("192.168.1.200").Equal(ip))
 	assert.Equal(t, uint8(0), key)
 }
```

</p>
</details> 

The behavior I'm seeing is that an entry with the weaker source is actually overwriting the existing entry. In this case, a different tunnel IP overwrites the existing one, which...doesn't seem good. I'm not sure if we've fundamentally missed something here in the merging logic. :worried:

joestringer replied:
I believe that's caused by the code in my comment below:
https://github.com/cilium/cilium/pull/23208#discussion_r1139431586

christarazi replied:
Ah, you're right. I also needed to ensure that the [above code](https://github.com/cilium/cilium/pull/23208#discussion_r1139431586) ran against the prefix infos that are sorted by source. If we don't sort by source, we get the wrong behavior as mentioned in my previous msg.

joestringer commented on pkg/ipcache/types.go:
```diff
@@ -176,5 +229,33 @@ func (s prefixInfo) logConflicts(scopedLog *logrus.Entry) {
+					logfields.TunnelPeer:            tunnelPeer.String(),
+					logfields.Resource:              tunnelPeerResourceID,
+					logfields.ConflictingTunnelPeer: info.tunnelPeer.String(),
+					logfields.ConflictingResource:   resourceID,
```
I'm mildly suspicious that we may end up receiving this same information two ways in some case, from both k8s and etcd. Hmm. I don't think we have any testing with etcd enabled that would provide us feedback on this.

joestringer replied:
^ I idly wonder whether we should be sorting first by the source corresponding to the info, then by the resource, so that we reliably pick up the "most important" source. But I guess the point is moot if we don't receive tunnel mappings from multiple sources.

christarazi replied:
Ended up implementing `sortedByResourceIDsAndSource()`. That will make the logic select the info from the strongest source, i.e. kvstore info over k8s, if that case is ever hit.

See https://github.com/cilium/cilium/pull/23208#discussion_r1148791817.

joestringer commented on pkg/ipcache/types.go:
```diff
@@ -176,5 +229,33 @@ func (s prefixInfo) logConflicts(scopedLog *logrus.Entry) {
+					logfields.Key:                 encryptKey.String(),
+					logfields.Resource:            encryptKeyResourceID,
+					logfields.ConflictingKey:      info.encryptKey.String(),
+					logfields.ConflictingResource: resourceID,
```
I presume that in the case of key rotation, the Resource should be the same each time so we'll replace the old encryptKey info with the new one and all should be well? (I recognize that there's tests for this, but it assumes the same custom resource name for the source of the info, so I'm just double-checking that assumption)

christarazi replied:
Yes, the old prefix info will get overwritten as each node have their own unique resource ID.

joestringer commented on pkg/ipcache/types.go:
```diff
@@ -176,5 +229,33 @@ func (s prefixInfo) logConflicts(scopedLog *logrus.Entry) {
+					"This may cause connectivity issues for this address.")
+			}
+			tunnelPeer = info.tunnelPeer
+			tunnelPeerResourceID = resourceID
```
Between the warning and the field setting here, I don't think it will be obvious to users which tunnel Peer is effective. I'd be inclined to interpret the one with "TunnelPeer" as the effective one, and the "Conflicting" one to be ignored. But it looks like the code is overwriting for each loop through the resources, so in the end it's the "Conflicting" one that ends up being used.

Maybe we can clarify this a bit or put this update piece in an `} else {` condition?

christarazi replied:
Implemented the update case with `else {`. From my read of the code, I think this was just an omission by mistake from the initial implementation.

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -354,44 +356,53 @@ func (m *manager) nodeAddressHasEncryptKey(address nodeTypes.Address) bool {
+		log.WithFields(logrus.Fields{
+			logfields.Source:   n.Source,
+			logfields.NodeName: n.Name,
+		}).Warning("received node with invalid node IPv4")
```
I welcome this additional checking here, though I also wonder if this function should just return if the nodeIP is invalid - it seems weird to attempt to process the rest of the update if we clearly can't support this node based on its primary IP.

(Separately I wonder whether we can just change the `n.GetNodeIP(...)` call here to pick up an IPv6 address if available, but that's digressing from the point of this PR, we can add IPv6 node support separately)

joestringer replied:
^ More discussion in other threads below but I'd encourage ignoring the update (and maybe make sure the log message here makes that clear). Plus a similar corresponding check on delete so we don't attempt to delete a node with an invalid IPv4 address.

christarazi replied:
I removed this code as it provided no extra benefit. I went back and checked all possible calls into NodeUpdated(). Either it comes from the K8s watcher or from the etcd, both of which should have parsed the IP correctly, so it's impossible for NodeUpdated to receive a malformed address (as of how this code works today). Certainly, this can change going forward, but no need for this PR to add defensive checks.

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -354,44 +356,53 @@ func (m *manager) nodeAddressHasEncryptKey(address nodeTypes.Address) bool {
+	return m.legacyNodeIpBehavior() && address.Type != addressing.NodeCiliumInternalIP
+}
+
+func nodeIdenitityLabels(n nodeTypes.Node) (nodeLabels labels.Labels, hasOverride bool) {
```
nit: `nodeIdentityLabels`, not `nodeIdenitityLabels`.

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -354,44 +356,53 @@ func (m *manager) nodeAddressHasEncryptKey(address nodeTypes.Address) bool {
 	}
 
-	var ipsAdded, healthIPsAdded, ingressIPsAdded []string
+	resource := ipcacheTypes.NewResourceID(ipcacheTypes.ResourceKindNode, "", n.Name)
```
nit: Huh, interesting, I think that this might actually be coming from `cn`, ie `ciliumnodes` rather than k8s `node` resources. I may have messed that up earlier.

Background: I've been trying to get these ResourceIDs to match k8s resources so that in future when we dump them via some CLI or even logs, you can clearly identify where this data came from and go directly to the k8s resource to compare and confirm that ipcache is in sync with it.

joestringer replied:
(Upon further inspection, I'm not sure we know the real source of the info at this point. `cn` is probably closer to reality in most cases, but if this can be called via etcd update code then we'd need the caller to pass the ResourceID in. I think that's the right longer term approach.)

joestringer replied:
(feel free to take a look but not a blocker)

christarazi replied:
Good point, I think we'll need to address this in a followup PR. Added it to https://github.com/cilium/cilium/issues/21142.

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -402,66 +413,45 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
-		} else {
-			healthIPsAdded = append(healthIPsAdded, addrStr)
-		}
+		m.ipcache.UpsertMetadata(healthIP, n.Source, resource,
```
I think that the health IP, being an in-cluster, "Pod-like" IP, should probably also be overridden. Otherwise we may end up selecting it with CIDR policy, which is unintentional.

joestringer replied:
Strictly speaking the bug I'm referring to above is probably not possible until we switch CIDR policy over to the new API, but it'd be nice to add the one extra override line in this PR since it's handling the rest of the related code here, then the future PR won't have to worry about (or forget about) this detail.

christarazi replied:
Discussed offline: health IPs are already excluded in `resolveIdentity()` but we do need to ensure that ingress IPs are also excluded. I've updated https://github.com/cilium/cilium/issues/21142 so we can followup.

christarazi replied:
https://github.com/cilium/cilium/pull/23208#issuecomment-1509337688

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -402,66 +413,45 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
-		} else {
-			ingressIPsAdded = append(ingressIPsAdded, addrStr)
-		}
+		m.ipcache.UpsertMetadata(ingressIP, n.Source, resource,
```
There's a curious discussion to be had regarding ingress here as well - Should the ingress IP be selectable via CIDR policy? If yes, this is good as-is. If no, we should also do an override here.

joestringer replied:
Given the way this worked before, we're probably currently assuming that the Ingress IP is not selectable via CIDR policy.

christarazi replied:
https://github.com/cilium/cilium/pull/23208#discussion_r1145710690

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -402,66 +413,45 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
-			NewSrc:      n.Source,
-		}
-		if err != nil && !errors.Is(err, overwriteErr) {
-			dpUpdate = false
```
I'm a little bit nervous that the new code doesn't have any of this corresponding logic. It's a bit terse to understand the implications but it hits us in one of the areas we are not as well covered in tests - etcd mode.

Part of this is that we have a weird split of information where this node manager logic informs the ipcache about some info, which may notify the datapath for other reasons, but then this node manager logic also directly calls the datapath to update other logic via `nh.NodeUpdate()` below.[^1]

My understanding is that in the case of two sources for Node information (K8s CRD `CiliumNode` and etcd entries), this code would previously rely on the IPCache update to decide whether to ignore this node event. Under the hood this decision would be made via `pkg/source`.

The previous code did two steps in one: Update the ipcache, and figure out if this node update event can be ignored or not. We've switched the first part over to the new API.. now I wonder if we should handle the second part with some lookup. Basically try to determine whether (a) we've handled an event for a node like this before, and (b) whether this event currently being handled has a higher priority source. Then we would set the `dpUpdate` bool accordingly.

[^1]:  This can mean that if we don't completely synchronize this code with IPCache's handling of the update, then we could end up in a situation where this code notifies the datapath to perform some update for some portion of the datapath, but then the IPCache skips the notification because it decides the update should not trigger updates to the datapath. Broadly this is a smell that the datapath abstraction and interaction with node manager and IPCache is not exactly consistent, but that's how it works today and I don't want to saddle this PR with the tech debt of solving that design problem.

christarazi replied:
Addressed in [diff](https://github.com/cilium/cilium/compare/9c2477eecf0200278577ff95a1550066525e9c2d..b1cdb35bbda3defca5117d97835a02a5c12048dd).

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -526,44 +481,83 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
+			continue
+		}
+
+		if option.Config.NodeIpsetNeeded() && address.Type == addressing.NodeInternalIP {
+			iptables.RemoveFromNodeIpset(address.IP)
+		}
```
This code is now shared with `NodeUpdated()`, which previously didn't do anything with ipset in the corresponding code... is this fixing a bug?

christarazi replied:
I'm confused, as the [update code](https://github.com/cilium/cilium/pull/23208/files#diff-e9cec4c2460e089f0f518c07f3d4c2fb40273486a8346d499ec5188395417772R397-R399) does have ipset logic.

joestringer replied:
I only see it in the `NodeDeleted()` logic in master: https://github.com/cilium/cilium/blob/4c6fd7e1e8a6fe5dc819c4d5e083c84b134d6861/pkg/node/manager/manager.go#L603

christarazi replied:
Maybe I'm misunderstanding what you mean by your original comment. There's both mirroring ipset logic in the NodeUpdated() and NodeDeleted() case. There isn't an asymmetry AFAICS.

joestringer replied:
:thinking: I'm saying that the current code in master doesn't do this, and now this patch *does* do this. That's a change in behaviour. If it's a change in behaviour, then why does the current code in master not need to clean up ipsets?

joestringer replied:
^ Either way the newest code looks correct, and we don't seem to be receiving bug reports about Cilium's node updates failing to delete ipset logic (at least that I've noticed), so maybe it's a bug on master but it's not impacting anyone.

christarazi replied:
Synced offline. The missing cleanup code from the NodeUpdated case on master has been addressed in https://github.com/cilium/cilium/pull/24825.

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -526,44 +481,83 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
+func (m *manager) removeNodeFromIPCache(oldNode nodeTypes.Node, resource ipcacheTypes.ResourceID,
+	nodeIPsAdded, healthIPsAdded, ingressIPsAdded []netip.Prefix) {
+
+	oldNodeIP, _ := ip.AddrFromIP(oldNode.GetNodeIP(false))
```
What's the case where the second `ok` bool returns false here? Is that possible?  Maybe related to my earlier discussion about skipping the event if there's no valid IPv4 address.

If we're skipping for `NodeUpdated()` then we should also skip for `NodeDeleted()` and then I think I'm ok with ignoring the second parameter here (ideally with a brief comment describing why -- that the callers already pre-check that there's a valid IPv4 address.)

christarazi replied:
https://github.com/cilium/cilium/pull/23208#discussion_r1148794708

joestringer reviewed (changes requested):
Apologies for the extended delay on this review :bow: 

I think there's a few things we should look a little more closely at, see comments below:
- CIDR policy is likely to be enforced on the health IP now, which is a behaviour change. Easy fix.
- There's some code that is being deleted which previously seemed to handle some cases where Nodes were backed by Etcd. I'm nervous that we don't have the coverage to point out failures there. I've got some suggestions that may allow us to preserve the current behaviour with the new API.
- A few other minor usability+visibility cleanups

christarazi committed:
ipcache: Rename GetIDMetadataByIP to GetMetadataLabelsByIP

The new name better represents what the function is. This commit has no
functional impact.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
ipcache: Define new GetMetadataByPrefix

This will be useful for fetching the metadata for particular entries in
the ipcache.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
ipcache: Export PrefixInfo

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/ipcache/types.go:
```diff
@@ -95,15 +114,24 @@ func (s prefixInfo) isValid() bool {
+	slices.Sort(resourceIDs)
+	slices.SortFunc(resourceIDs, func(a, b types.ResourceID) bool {
+		return !source.AllowOverwrite(s[a].source, s[b].source)
+	})
```
Hm, upon second thought, maybe we want to reverse this so that we get the strongest sources first to weakest source, but if any entries have matching sources, then those are sorted by resource ID (alphabetically). In other words, just swap these two lines around.

joestringer replied:
Yes, I agree.

joestringer commented on pkg/ipcache/types_test.go:
```diff
@@ -0,0 +1,41 @@
+	}
+	pi["generated-uid"] = &resourceInfo{
+		source: source.Generated,
+	}
```
Shouldn't one of these be `source.KVStore`?

joestringer commented on pkg/ipcache/types/types.go:
```diff
@@ -61,3 +63,38 @@ func NewResourceID(kind ResourceKind, namespace, name string) ResourceID {
+type TunnelPeer netip.Addr
+
+func (t TunnelPeer) IsValid() bool {
+	return netip.Addr(t).IsValid()
```
nit: I suspect if `type TunnelPeer` is defined as `struct { netip.Addr }`, it will inherit these functions and avoid the need for the wrappers?

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -332,134 +333,138 @@ func (m *manager) legacyNodeIpBehavior() bool {
+	return m.legacyNodeIpBehavior() && address.Type != addressing.NodeCiliumInternalIP
+}
+
+func nodeIdentityLabels(n nodeTypes.Node) (nodeLabels labels.Labels, hasOverride bool) {
```
I think this should maybe take into account `m.conf.RemoteNodeIdentitiesEnabled` like the earlier corresponding code did? Otherwise even with remote node identities disabled, it will set the labels to remote node and break connectivity for users with that setting.

joestringer replied:
Unfortunately I don't think the test coverage is likely to tell you much on this front :/ I just noticed the difference while comparing this patch just a few lines down in the next hunk (the code that got removed).

christarazi replied:
Good catch, addressed:

<details><summary>Details</summary>
<p>

```diff
commit 40ac0845b550f99fbfca5632b8d26541d5f0dcc3
Author: Chris Tarazi <chris@isovalent.com>
Date:   Tue Apr 11 22:03:28 2023 -0700

    fixup! nodemanager: Move to async IPCache API
    
    Signed-off-by: Chris Tarazi <chris@isovalent.com>

diff --git a/pkg/node/manager/manager.go b/pkg/node/manager/manager.go
index c341e8fdcb..f2d6ae18cc 100644
--- a/pkg/node/manager/manager.go
+++ b/pkg/node/manager/manager.go
@@ -359,14 +359,18 @@ func (m *manager) nodeAddressSkipsIPCache(address nodeTypes.Address) bool {
 	return m.legacyNodeIpBehavior() && address.Type != addressing.NodeCiliumInternalIP
 }
 
-func nodeIdentityLabels(n nodeTypes.Node) (nodeLabels labels.Labels, hasOverride bool) {
+func (m *manager) nodeIdentityLabels(n nodeTypes.Node) (nodeLabels labels.Labels, hasOverride bool) {
 	nodeLabels = labels.NewFrom(labels.LabelRemoteNode)
-	if n.IsLocal() {
+	if m.conf.RemoteNodeIdentitiesEnabled() {
+		if n.IsLocal() {
+			nodeLabels = labels.NewFrom(labels.LabelHost)
+		} else if !identity.NumericIdentity(n.NodeIdentity).IsReservedIdentity() {
+			// This needs to match clustermesh-apiserver's VMManager.AllocateNodeIdentity
+			nodeLabels = labels.Map2Labels(n.Labels, labels.LabelSourceK8s)
+			hasOverride = true
+		}
+	} else {
 		nodeLabels = labels.NewFrom(labels.LabelHost)
-	} else if !identity.NumericIdentity(n.NodeIdentity).IsReservedIdentity() {
-		// This needs to match clustermesh-apiserver's VMManager.AllocateNodeIdentity
-		nodeLabels = labels.Map2Labels(n.Labels, labels.LabelSourceK8s)
-		hasOverride = true
 	}
 	return nodeLabels, hasOverride
 }
@@ -390,7 +394,7 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
 	}
 
 	resource := ipcacheTypes.NewResourceID(ipcacheTypes.ResourceKindNode, "", n.Name)
-	nodeLabels, nodeIdentityOverride := nodeIdentityLabels(n)
+	nodeLabels, nodeIdentityOverride := m.nodeIdentityLabels(n)
 
 	var nodeIPsAdded, healthIPsAdded, ingressIPsAdded []netip.Prefix
 
@@ -519,7 +523,7 @@ func (m *manager) removeNodeFromIPCache(oldNode nodeTypes.Node, resource ipcache
 		// See comment in NodeUpdated().
 		oldNodeIP, _ = ip.AddrFromIP(nIP)
 	}
-	oldNodeLabels, oldNodeIdentityOverride := nodeIdentityLabels(oldNode)
+	oldNodeLabels, oldNodeIdentityOverride := m.nodeIdentityLabels(oldNode)
 
 	// Delete the old node IP addresses if they have changed in this node.
 	for _, address := range oldNode.IPAddresses {

```

</p>
</details>

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -332,134 +333,138 @@ func (m *manager) legacyNodeIpBehavior() bool {
+		// the IP is valid as long as it's coming from nodeTypes.Node. This
+		// object is created either from the node discovery (K8s) or from an
+		// event from the kvstore.
+		nodeIP, _ = ip.AddrFromIP(nIP)
```
Maybe worth filing an issue to support IPv6-only nodes and add `GH-XXXX:` here as a reference?

christarazi replied:
https://github.com/cilium/cilium/issues/24829

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -332,134 +333,138 @@ func (m *manager) legacyNodeIpBehavior() bool {
-			ipsAdded = append(ipsAdded, ipAddrStr)
+			continue
+		}
+		m.ipcache.UpsertMetadata(prefix, n.Source, resource,
```
`dpUpdate=false` here makes sense.

I wonder if we should be Upserting the metadata anyway in this case though, this will quietly hide the updates from the lower priority source.

I don't know what the ordering guarantees are here, but let's say we learn about the same node from k8s and from kvstore. Depending on whether we handle the event for k8s then kvstore, or kvstore then k8s, the ipcache will either end up with two entries for the same node or one entry for the same node. Seems a bit inconsistent?

But then again, I can't think of a concrete example of how this could occur and then lead to some weird follow-on state, it's just that for instance if we could query all internal ipcache state then we might not necessarily see one of the pieces of information used to make the decision about the node events. If we did happen to have a bug in the logic, it could plausibly lead to some wrong behaviour but that's a bit hard to speculate about.

The only other case I could think of is if we ended up supporting scale up + scale down and dynamically enabling+disbaling kvstore. We have no such plans today.

:shrug:

joestringer replied:
(Even if we upserted the metadata from both cases, the nodeIdentityOverride below should only occur iff `dpupdate==true`.)

joestringer replied:
> Depending on whether we handle the event for k8s then kvstore, or kvstore then k8s, the ipcache will either end up with two entries for the same node or one entry for the same node.

Actually this is not quite correct, I was assuming unique `ResourceID`s for the resource sourced from etcd vs. the one sourced from k8s. That's how I think it _should_ work, so we can trace the true source of the data for the ipcache entry back to where it came from. But in practice the ResourceID is set directly earlier in this function for both sources, so there's no way to tell where it came from. It also means that the ipcache entries would be updated / overwritten by the update from the other source.

We also discussed in person previously that we assume that both sources are going to end up propagating the exact same information anyway, so this whole line of consideration is probably more theoretical than practical. I think it's probably fine to chalk this up as "feels a little odd but probably not going to have any consequences" and let my concerns lay to rest :sweat_smile:

christarazi replied:
As a followup, I've [wrote down](https://github.com/cilium/cilium/issues/21142) to come back and rework the resource IDs here so that we can capture kvstore vs k8s.

joestringer reviewed (approved):
Down to the long tail of low-severity nits now I think. Nothing particular is popping out from review this time :+1:

gandro commented on pkg/ipcache/metadata.go:
```diff
@@ -125,19 +125,27 @@ func (m *metadata) upsertLocked(prefix netip.Prefix, src source.Source, resource
+	m := ipc.metadata.getLocked(prefix)
+	n := make(prefixInfo, len(m))
+	for k, v := range m {
+		n[k] = v
```
`v` is a pointer of type `*resourceInfo` which can be mutated in place e.g. here:

https://github.com/cilium/cilium/blob/1c01d518141b011ab75f398a350b4ead7c6b316b/pkg/ipcache/metadata.go#L121

I think we need to create a copy of `v` here, otherwise we can have mutation happen on it while we are reading from it - or alternatively only return the `Source()` part here.

christarazi replied:
Good catch, addressed

<details><summary>Details</summary>
<p>

```diff
commit 446e180d3e9e58336b5425e3bb0a056a227e1083
Author: Chris Tarazi <chris@isovalent.com>
Date:   Wed Apr 12 16:04:48 2023 -0700

    fixup! ipcache: Define new GetMetadataByPrefix
    
    Signed-off-by: Chris Tarazi <chris@isovalent.com>

diff --git a/pkg/ipcache/metadata.go b/pkg/ipcache/metadata.go
index c631d9f433..84dfd37112 100644
--- a/pkg/ipcache/metadata.go
+++ b/pkg/ipcache/metadata.go
@@ -141,7 +141,7 @@ func (ipc *IPCache) GetMetadataByPrefix(prefix netip.Prefix) PrefixInfo {
 	m := ipc.metadata.getLocked(prefix)
 	n := make(PrefixInfo, len(m))
 	for k, v := range m {
-		n[k] = v
+		n[k] = v.DeepCopy()
 	}
 	return n
 }
diff --git a/pkg/ipcache/types.go b/pkg/ipcache/types.go
index fc7d7709c7..f31385962c 100644
--- a/pkg/ipcache/types.go
+++ b/pkg/ipcache/types.go
@@ -21,6 +21,9 @@ import (
 // independently based on the ResourceID of the origin of that information, and
 // provides convenient accessors to consistently merge the stored information
 // to generate ipcache output based on a range of inputs.
+//
+// Note that when making a copy of this object, resourceInfo is pointer which
+// means it needs to be deep-copied via (*resourceInfo).DeepCopy().
 type PrefixInfo map[types.ResourceID]*resourceInfo
 
 // IdentityOverride can be used to override the identity of a given prefix.
@@ -105,6 +108,16 @@ func (m *resourceInfo) isValid() bool {
 	return false
 }
 
+func (m *resourceInfo) DeepCopy() *resourceInfo {
+	n := new(resourceInfo)
+	n.labels = labels.NewFrom(m.labels)
+	n.source = m.source
+	n.identityOverride = m.identityOverride
+	n.tunnelPeer = m.tunnelPeer
+	n.encryptKey = m.encryptKey
+	return n
+}
+
 func (s PrefixInfo) isValid() bool {
 	for _, v := range s {
 		if v.isValid() {
```

</p>
</details>

gandro reviewed (commented):
Thanks a lot for pushing it over the finish line. Only one thing that stood out to me which looks like a data race.

christarazi committed:
ipcache: Exclude ingress reserved identity from CIDR identity

Similar to the health reserved identity, the ingress identity shouldn't
have CIDR labels associated with it, so exclude it from the identity
resolution logic. The ingress IPs comes from the CiliumNode object.

Without this commit, ingress IPs will have a CIDR labels and therefore a
CIDR identity, instead of having the reserved ingress identity.

Related: https://github.com/cilium/cilium/issues/21142

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
node/manager: Refactor skip ipcache helper function

This pulls out logic that is used in both NodeUpdated() and
NodeDeleted() for proper code reuse.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/ipcache/metadata.go:
```diff
@@ -218,27 +236,31 @@ func (ipc *IPCache) InjectLabels(ctx context.Context, modifiedPrefixes []netip.P
 				break
 			}
 
-			// We can safely skip the ipcache upsert if the ID and source
-			// in the metadata cache match the ipcache exactly.
-			// Note that checking ID alone is insufficient, see GH-24502
-			if oldID.ID == newID.ID && prefixInfo.Source() == oldID.Source {
+			// We can safely skip the ipcache upsert if the entry matches with
+			// the entry in the metadata cache exactly.
+			// Note that checking ID alone is insufficient, see GH-24502.
+			if oldID.ID == newID.ID && prefixInfo.Source() == oldID.Source &&
+				oldTunnelIP.Equal(prefixInfo.TunnelPeer().IP()) &&
+				oldEncryptionKey == prefixInfo.EncryptKey().Uint8() {
 				goto releaseIdentity
 			}
```
Pointing this out as a key diff from previous revision of the PR

christarazi replied:
cc @gandro @joestringer, this was needed to resolve the encryption issue on upgrade on kind clusters (with tunneling), part 1/2.

gandro replied:
Good catch! Should we add a `prefixInfo.HasChanged(other)` method that performs this the check on auxiliary (i.e. metadata that does not change the ID)? We might add more non-identity defining metadata in the future (e.g. named ports) that will very likely need the same check here.

christarazi replied:
Yes good idea. I'll add it to https://github.com/cilium/cilium/issues/21142.

christarazi commented on pkg/node/manager/manager.go:
```diff
@@ -333,134 +333,153 @@ func (m *manager) legacyNodeIpBehavior() bool {
+		prefix := ip.IPToNetPrefix(address.IP)
+		// We expect the node manager to have a source of either Kubernetes,
+		// CustomResource, or KVStore. Prioritize the KVStore source over the
+		// rest as it is the strongest source, i.e. only trigger datapath
+		// updates if the information we receive takes priority.
+		//
 		// The only exception are kube-apiserver entries. In that case,
 		// we still want to inform subscribers about changes in auxiliary
 		// data such as for example the health endpoint.
-		overwriteErr := &ipcache.ErrOverwrite{
-			ExistingSrc: source.KubeAPIServer,
-			NewSrc:      n.Source,
-		}
-		if err != nil && !errors.Is(err, overwriteErr) {
+		existing := m.ipcache.GetMetadataByPrefix(prefix).Source()
+		overwrite := source.AllowOverwrite(existing, n.Source)
+		if !overwrite && existing != source.KubeAPIServer {
```
Ultimately, checking for `existing != source.KubeAPIServer` was missing from the previous revision due to incorrect rebase. This contributed to CI kind failures.

christarazi replied:
cc @gandro @joestringer, this was needed to resolve the encryption issue on upgrade on kind clusters (with tunneling), part 2/2.

gandro replied:
Wrote down my thoughts on this here: https://github.com/cilium/cilium/pull/23208#issuecomment-1543570384

I don't think we should address this in this PR (it's already hard enough keeping track of the changes), but I do think we should think about this as part of the refactor, because I really think kube-apiserver is not that special. Future sources (e.g. label-sources) might suffer a similar problem.

christarazi replied:
Added it to https://github.com/cilium/cilium/issues/21142, agreed.

michi-covalent reviewed (approved):
reviewed logfields.go. looks innocent 😇

asauber commented on pkg/ipcache/ipcache.go:
```diff
@@ -225,6 +225,12 @@ func (ipc *IPCache) GetHostIP(ip string) net.IP {
 
+func (ipc *IPCache) GetHostIPCache(ip string) (net.IP, uint8) {
+	ipc.mutex.RLock()
+	defer ipc.mutex.RUnlock()
```
This doesn't guarantee that the caller is able to obtain the lock, so why perform this lock/unlock here?

asauber replied:
Ah nevermind, based on the name of the functions it looked like this was returning a handle to the entire cache. This is definitely nitpicky, but I would call these `GetHostCachedIP` and `getHostCachedIP`

asauber reviewed (approved):
Approving config-related changes
